src/itouch.cpp

```diff
@@ -0,0 +1,399 @@
+
+    if (getRodsEnv(&env) < 0) {
+        std::cerr << "Error: Could not get iRODS environment.\n";
+        return 1;
+    }
+
+    po::options_description desc{""};
+    desc.add_options()
+        ("no-create,c", "")
+        ("replica,n", po::value<replica_number_type>(), "")
+        ("resource,R", po::value<std::string>(), "")
+        ("reference,r", po::value<std::string>(), "")
+        ("seconds-since-epoch,s", po::value<rep_type>(), "")
+        ("logical_path", po::value<std::string>(), "")
+        ("help,h", "")
+        ("version,v", "");
+
+    po::positional_options_description pod;
+    pod.add("logical_path", 1);
+
+    try {
+        po::variables_map vm;
+        po::store(po::command_line_parser(_argc, _argv).options(desc).positional(pod).run(), vm);
+        po::notify(vm);
+
+        if (vm.count("help")) {
+           print_usage_info(); 
+           return 0;
+        }
+
+        if (vm.count("version")) {
+            print_version_info();
+            return 0;
+        }
+
+        if (!is_input_valid(vm)) {
+            return 1;
+        }
+
+        load_client_api_plugins();
+
+        std::string path;
+
+        if (auto v = canonical(vm["logical_path"].as<std::string>(), env); v) {
+            path = *v;
+        }
+        else {
+            std::cerr << "Error: Failed to convert path to absolute path.\n";
+            return 1;
+        }
+
+        ix::client_connection conn;
+
+        create_data_object_if_necessary(conn, env, vm, path);
+
+        //
+        // The target object exists. Update the mtime based on the arguments passed by the user.
+        //
+
+        const auto object_status = fs::client::status(conn, path);
+
+        if (!fs::client::is_collection(object_status) && !fs::client::is_data_object(object_status)) {
+            std::cerr << "Error: Logical path does not point to a collection or data object.\n";
+            return 1;
+        }
+
+        const auto new_mtime = get_time(conn, env, vm);
+
+        if (fs::client::is_collection(object_status)) {
+            // Return immediately if invalid options are present in regards to the target.
+            if (vm.count("replica") || vm.count("resource")) {
+                std::cerr << "Error: -n and -R cannot be used for collections.\n";
+                return 1;
+            }
+
+            if (fs::client::is_collection_registered(conn, path)) {
+                fs::client::last_write_time(conn, path, new_mtime);
+            }
+            else {
+                // TODO Should anything happen here?
+            }
+        }
+        else if (fs::client::is_data_object(object_status)) {
+            if (fs::client::is_data_object_registered(conn, path)) {
+                const auto replica_number = get_replica_number(vm, conn, path);
+                ix::replica::last_write_time<RcComm>(conn, path, replica_number, new_mtime);
+            }
+            else {
+                // TODO Should anything happen here?
```

Comment:
and this is a data object that doesn't exist in the catalog?   part of a special collection archive file?

Reply:
yes, it doesn't exist in the catalog.  likely part of a special collection

Reply:
maybe we print a message saying the mtime cannot be updated.

Reply:
seems `LOG_NOTICE` is good here... it's not a warning, unless we're asserting they shouldn't have gotten into this codepath.